### PR TITLE
Merges upstream changes from swift-DEVELOPMENT-SNAPSHOT-2018-07-23-a into tensorflow

### PR DESCRIPTION
**Pull request by bgogul**

### CODE REVIEW

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

File: lib/Sema/CSGen.cpp [hunk 1]
```diff
@@ -1505,7 +1505,7 @@ namespace {
           assert(adjointDecl->isInstanceMember() || adjointDecl->isStatic() &&
                  "Expected adjoint to be an instance/static method");
           auto methodContext = CurDC->getInnermostMethodContext();
-          baseType = methodContext->getParameterList(0)->get(0)->getType();
+          baseType = methodContext->getParameters()->get(0)->getType();
```
Comment (slavapestov) [inline lib/Sema/CSGen.cpp]: getParameterList(0) was the 'self' parameter list. I think you want `methodContext->getImplicitSelfDecl()->getType()`.

File: lib/Sema/DerivedConformanceParameterAggregate.cpp [hunk 2]
```diff
@@ -232,7 +231,7 @@ static ValueDecl *deriveParameterAggregate_update(DerivedConformance &derived) {
   }
   updateDecl->setInterfaceType(interfaceType);
```
Comment (slavapestov) [inline lib/Sema/DerivedConformanceParameterAggregate.cpp]: Instead, please just call `updateDecl->computeType()`, and remove the code for building your own `FunctionType`.

File: lib/Sema/TypeCheckAttr.cpp [hunk 3]
```diff
@@ -2345,7 +2345,7 @@ void AttributeChecker::visitDifferentiableAttr(DifferentiableAttr *attr) {
       //   function's result type.
       auto primalSelfDecl = primalCandidate->getImplicitSelfDecl();
```
Comment (slavapestov) [inline lib/Sema/TypeCheckAttr.cpp]: Unused variable?

File: lib/Sema/CSGen.cpp (hunk 1 above)
Comment (rxwei) [reply lib/Sema/CSGen.cpp]: This was done here: https://github.com/apple/swift/commit/652b33b51dea88ce3fa190b43db2adb55221b082.

@bgogul, was this overridden by rebase somehow?

File: lib/Sema/DerivedConformanceParameterAggregate.cpp (hunk 2 above)
Comment (rxwei) [reply lib/Sema/DerivedConformanceParameterAggregate.cpp]: @dan-zheng Please take a look. Thanks!

File: lib/Sema/CSGen.cpp (hunk 1 above)
Comment (rxwei) [reply lib/Sema/CSGen.cpp]: This was done here: https://github.com/apple/swift/commit/652b33b51dea88ce3fa190b43db2adb55221b082.

@bgogul, was this overridden by rebase somehow?

File: lib/Sema/DerivedConformanceParameterAggregate.cpp (hunk 2 above)
Comment (rxwei) [reply lib/Sema/DerivedConformanceParameterAggregate.cpp]: @dan-zheng please take a look. thanks!

File: lib/Sema/CSGen.cpp (hunk 1 above)
Comment (bgogul) [reply lib/Sema/CSGen.cpp]: This commit is included in the merge. If you change the diff view include changes from all commits, you will see this is fixed.

**Pull request merged**